plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,454 @@
+package jenkins
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/kelwang/gojenkins"
+)
+
+// Jenkins plugin gathers information about the nodes and jobs running in a jenkins instance.
+type Jenkins struct {
+	URL      string
+	Username string
+	Password string
+	// HTTP Timeout specified as a string - 3s, 1m, 1h
+	ResponseTimeout internal.Duration
+
+	tls.ClientConfig
+	instance *gojenkins.Jenkins
+
+	MaxConnections    int               `toml:"max_connections"`
+	MaxBuildAge       internal.Duration `toml:"max_build_age"`
+	MaxSubJobDepth    int               `toml:"max_subjob_depth"`
+	MaxSubJobPerLayer int               `toml:"max_subjob_per_layer"`
+	JobExclude        []string          `toml:"job_exclude"`
+	jobFilter         filter.Filter
+
+	NodeExclude []string `toml:"node_exclude"`
+	nodeFilter  filter.Filter
+}
+
+type byBuildNumber []gojenkins.JobBuild
+
+const sampleConfig = `
+url = "http://my-jenkins-instance:8080"
+#  username = "admin"
+#  password = "admin"
+## Set response_timeout
+response_timeout = "5s"
+
+## Optional SSL Config
+#  ssl_ca = /path/to/cafile
+#  ssl_cert = /path/to/certfile
+#  ssl_key = /path/to/keyfile
+## Use SSL but skip chain & host verification
+#  insecure_skip_verify = false
+
+## Job & build filter
+#  max_build_age = "1h"
+## jenkins can have unlimited layer of sub jobs
+## this config will limit the layers of pull, default value 0 means
+## unlimited pulling until no more sub jobs
+#  max_subjob_depth = 0
+## in workflow-multibranch-plugin, each branch will be created as a sub job
+## this config will limit to call only the lasted branches
+## sub jobs fetch in each layer
+#  empty will use default value 10
+#  max_subjob_per_layer = 10
+#  job_exclude = [ "job1", "job2/subjob1/subjob2", "job3/*"]
+
+## Node filter
+#  node_exclude = [ "node1", "node2" ]
+
+## Woker pool for jenkins plugin only
+#  empty this field will use default value 30
+#  max_connections = 30
+`
+
+// measurement
+const (
+	measurementNode = "jenkins_node"
+	measurementJob  = "jenkins_job"
+)
+
+// Error base type of error.
+type Error struct {
```

Comment:
Would you remove this Error struct and put the errors inline in the code?

Reply:
removed

Reply:
👍